codegen-v2/src/main/kotlin/club/doki7/babel/cdecl/parse.kt

```diff
@@ -1,11 +1,424 @@
+        val trivia = mutableListOf<String>()
+        trivia.addAll(getAndSkipTrivia())
+        expect(TokenType.IDENT)
+        val name = curToken.value
+        pos++
+        trivia.addAll(getAndSkipTrivia())
+        expect(TokenType.SYMBOL, "=")
+        pos++
+        trivia.addAll(getAndSkipTrivia())
+        if (hasEnded || curToken.type != TokenType.INTEGER && curToken.type != TokenType.IDENT) {
+            throw ParseException("expected integer or identifier", curToken)
+        }
+        val value = curToken.value
+        pos++
+        trivia.addAll(getAndSkipTrivia())
+        if (hasEnded || curToken.type != TokenType.SYMBOL || curToken.value != ",") {
+            throw ParseException("expected ','", curToken)
+        }
+        return EnumeratorDecl(
+            name = name,
+            value = value,
+            syntaxTrivia = trivia
+        )
+    }
+}
+
+fun parseStructFieldDecl(source: List<String>, startLine: Int): Pair<StructFieldDecl, Int> {
+    val (tokens, lastLine) = CTokenizer(source, startLine).tokenize(',')
+    val (name, ty) = CTypeParser(tokens).parseDeclaration()
+    return Pair(
+        StructFieldDecl(
+            fieldName = name,
+            fieldType = ty,
+            syntaxTrivia = emptyList()
+        ),
+        lastLine
+    )
+}
+
+fun parseFunctionDecl(source: List<String>, startLine: Int): Pair<FunctionDecl, Int> {
+    val (tokens, lastLine) = CTokenizer(source, startLine).tokenize(';')
+    val (name, ty) = CTypeParser(tokens).parseDeclaration()
+    return Pair(
+        FunctionDecl(
+            functionName = name,
+            returnType = ty,
+            params = emptyList(),
+            syntaxTrivia = emptyList()
+        ),
+        lastLine
+    )
+}
+
+fun parseFunctionTypedefDecl(source: List<String>, startLine: Int): Pair<FunctionTypedefDecl, Int> {
+    val (tokens, lastLine) = CTokenizer(source, startLine).tokenize(';')
+    val (name, ty) = CTypeParser(tokens).parseTypedefDecl()
+    assert(ty is RawFunctionType) { "typedef expects function type" }
+    val funcTy = ty as RawFunctionType
+    return Pair(
+        FunctionTypedefDecl(
+            functionName = name,
+            returnType = funcTy.returnType,
+            params = funcTy.params.map { FunctionParamDecl(it.second, it.first, emptyList()) },
+            syntaxTrivia = funcTy.syntaxTrivia
+        ),
+        lastLine
+    )
+}
+
+fun parseTypedefDecl(source: List<String>, startLine: Int): Pair<TypedefDecl, Int> {
+    val (tokens, lastLine) = CTokenizer(source, startLine).tokenize(';')
+    val (name, ty) = CTypeParser(tokens).parseTypedefDecl()
+    assert(ty !is RawFunctionType) { "typedef expects non-function type" }
+    return Pair(
+        TypedefDecl(
+            typeName = name,
+            aliasedType = ty,
+            syntaxTrivia = emptyList()
+        ),
+        lastLine
+    )
+}
+
+fun parseEnumeratorDecl(source: List<String>, startLine: Int): Pair<EnumeratorDecl, Int> {
```

Comment:
Comma (`,`) is not the only `stopAtChar` for enumerator declaration:

```c
enum {
    A = 1124,
    B = 1125 // <-
};
```

Reply:
you can change it to use EOF or comma

Reply:
But as far as Chuigda's API is concerned (return a line number where the decl ends), the point is that caller actually doesn't know where for `parseEnumeratorDecl` to stop parsing. Here the stop character should be '}' instead. I'll take this.